### DISCUSSION TITLE
Why doesn't adapter-node bundle node_modules?

### DISCUSSION THREAD

**Discussion by parischap** [sveltejs/kit#2822] [Q&A] [answered]
When I build my sveltekit project with adapter-node, I realize that the contents of the node_modules folder is not bundled into the build directory. So when I upload my production website onto a server, I have to install all my node_modules there as well. I analyzed the code of the node adapter and discovered this is on purpose. On line 66 of the main index.js file, we see:

`external: Object.keys(JSON.parse(readFileSync('package.json', 'utf8')).dependencies || {}),
`
Why is that ?

**Comment by bluwy** [comment, marked as answer]
It's because there's no need to. In a normal node environment, it's expected for prod dependencies to be installed, and the final build should leverage that instead of bundling them. The reason why other adapters like netlify or cloudflare does is mainly that they are in a worker environment and it's expected for everything to be bundled and be as small as possible.

**Comment by parischap** [comment]
Thanks for your answer. What do you mean by 'the final build should leverage that'? Are there any benefits in not bundling the node_modules folder in the final build? In other words, if I decide to bundle the node_modules folder (because I have space restriction on the server I use), do I have any drawbacks?

**Reply by bluwy** [reply]
> What do you mean by 'the final build should leverage that'?

I meant that the final build would have js code that imports packages from the node_module, which is the default behaviour in adapter node.

Re bundle vs not bundle, the main difference on top of my head is that bundling would take more time to build due to the need to parse more files.

Another reason to not bundle could be if you want to change env vars without the need to rebuild.

Perhaps these choices would come down to personal preference, but I would lean to default node behaviour whenever possible.

**Comment by loxs** [comment]
One reason to bundle is because in Docker land an image of 1.5G is considered enormous.

**Comment by dmgarland** [comment]
Doesn't that mean that we lose the benefits of tree shaking in the production dependencies?

**Comment by hahouari** [comment]
Would really love for this feature to get some attention especially for Docker users, at least to provide some workarounds, here is a deployment example of our deployment having SurrealDB and an Actix Web server in Rust both are lightweight (less than `100MB`), but the sveltekit app `tojema_homepage` is insanly taking `~320MB` just for a SvelteKit SPA.

![image](https://github.com/sveltejs/kit/assets/39862612/1a8c1e5c-68ae-4a58-aa96-e64b05e0ff83)

If anyone can suggest a way to work around this, would be much appreciated.
